### ISSUE TITLE
Can't use App.xaml resources after Dependency Injection .NET MAUI

### ISSUE THREAD

**Issue by luis95gr** [dotnet/maui#11485]
### Description

I have an application that uses mvvm pattern with Community Toolkit features. I am trying to use Dependency Injection to use an ApiService with its interface in my viewmodels, but after following the steps described [here](https://stackoverflow.com/questions/72429297/injecting-services-into-view-models-in-net-maui-app), I can't access to App.xaml Resources (specifically colors), the intellisense works when I am writing code in XAML, but it doesn´t work after running. It is important to notice that I was using colors from resources correctly before trying to use Dependency Injection and changing the ViewModel - View linking method to the one described [here](https://stackoverflow.com/questions/72429297/injecting-services-into-view-models-in-net-maui-app), but I was unable to use ApiService. Here is my code.

### Steps to Reproduce

App.xaml.cs (Login page is my first page):
```
public App(LoginPage page)
{
    InitializeComponent();
    MainPage = page;
}
```
LoginPage.xaml.cs
```
public partial class LoginPage : ContentPage
{
  public LoginPage(LoginPageViewModel loginPageViewModel)
  {
    BindingContext = loginPageViewModel;
    InitializeComponent();
  }
}
```
LoginPageViewModel
```
public LoginPageViewModel(IApiService apiService)
    {
        _apiService = apiService;
        Opacity = 1f;
        IsEnabled = true;
        IsRunning = false;
    }
```
LoginPage.xaml
```
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
         xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
         x:Class="NewScholarApp.Views.LoginPage"
         Title="LoginPage"
         xmlns:vm="clr-namespace:NewScholarApp.ViewModels"
         x:DataType="vm:LoginPageViewModel"
         BackgroundColor="{StaticResource GreenSchool}">
```
RegisterViews and ViewModels
```
public static MauiAppBuilder ConfigureViews(this MauiAppBuilder mauiAppBuilder)
    {
        mauiAppBuilder.Services.AddTransient<LoginPage>();

        return mauiAppBuilder;
    }

public static class RegisterViewModels
{

    public static MauiAppBuilder ConfigureViewModels(this MauiAppBuilder mauiAppBuilder)

    {
        mauiAppBuilder.Services.AddSingleton<LoginPageViewModel>();

        return mauiAppBuilder;
    }
}
```

### Link to public reproduction project repository

https://github.com/luis95gr/NewScholarApp

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android, I was *not* able test on other platforms

### Affected platform versions

Android 12.1

### Did you find any workaround?

Don't use any StaticResource

### Relevant log output

```shell
Microsoft.Maui.Controls.Xaml.XamlParseException: 'Position 8:14. StaticResource not found for key GreenSchool'
```

**Comment by TWTIC** [comment]
Try to change the order like this:

public partial class LoginPage : ContentPage
{
  public LoginPage(LoginPageViewModel loginPageViewModel)
  { 
    InitializeComponent();
    BindingContext = loginPageViewModel;
  }
}

**Comment by luis95gr** [comment]
> Try to change the order like this:
> 
> public partial class LoginPage : ContentPage { public LoginPage(LoginPageViewModel loginPageViewModel) { InitializeComponent(); BindingContext = loginPageViewModel; } }

Thank you for the response, but the error still happens even with that change.

**Comment by symbiogenesis** [comment]
I encountered this same issue when upgrading to MAUI and switching to the new DI style.

I didn't want to change all my StaticResource bindings to DynamicResource, since that incurs a cost.

The two options for workarounds I found are the following:

1. For the App's MainPage, inject only the ViewModel, not the page, and just new up the page.
2. Inject the main page using System.Lazy as described [here](https://dev.to/hypercodeplace/lazy-dependency-injection-37en) or [here](https://stackoverflow.com/questions/44934511/does-net-core-dependency-injection-support-lazyt)

**Comment by Eilon** [comment]
@luis95gr - we don't seem to have access to https://github.com/luis95gr/NewScholarApp. Can you ensure the repo is public?

**Comment by luis95gr** [comment]
> @luis95gr - we don't seem to have access to https://github.com/luis95gr/NewScholarApp. Can you ensure the repo is public?

I just update the original post with a public repo. This is the link: [https://github.com/luis95gr/SampleProject](url)

**Comment by satya-prismhr** [comment]
Hi @luis95gr I have a workaround for it, you can create service helper class with this code

public static class ServiceHelper
{
    public static TService GetService<TService>()
        => Current.GetService<TService>();

    public static IServiceProvider Current =>
#if WINDOWS10_0_17763_0_OR_GREATER
			MauiWinUIApplication.Current.Services;
#elif ANDROID
            MauiApplication.Current.Services;
#elif IOS || MACCATALYST
			MauiUIApplicationDelegate.Current.Services;
#else
			null;
#endif
}


and use it where ever it required
        MyViewModel myViewModel = ServiceHelper.Current.GetService<MyViewModel >();

Do not register pages in builder.Services, just register services and Viewmodels

**Comment by symbiogenesis** [comment]
If you use DI for everything, you can just use constructor injection to resolve IServiceProvider from anywhere in your application, and you shouldn't need a static IServiceProvider exposed.

Except maybe for integration testing purposes, and then you can use an ifdef for that case.

**Comment by satya-prismhr** [comment]
Hi @symbiogenesis 

Thanks for the input, Please share if you have any working sample project with IServiceProvider implementation in MAUI. It will help many .NET folks

Thanks in advance

**Comment by symbiogenesis** [comment]
Modified

https://github.com/symbiogenesis/ServiceProviderInjection

**Comment by symbiogenesis** [comment]
What is happening is that if you use DI to register ContentPage elements and maybe other UI elements, you will not be able to access static resources declared in your App.xaml file. I'm sure the technical reasons involve the app lifecycle assumptions deep within MAUI. A UI element resolved via constructor injection is initialized very early in the app or page lifecycle, and it can only access those resources later on. I believe I saw other bugs also logged about this, and maybe it will get fixed.

But essentially you can get around this by injecting IServiceProvider into the App class, and then using that to resolve the MainPage inside the App class. forcing the registration of the page element to occur later than the constructor injection, which I demonstrated in the comment above yours.

**Comment by janseris** [comment]
@symbiogenesis  
Thanks a lot for pointing the root cause out.
I did not know that actually the instance is lazily created when you call GetService/GetRequiredService, I assumed it is already created anyways.
Injecting IServiceProvider really helped solving the issue.
My use case was:

Injecting AppShell into App because AppShell needs injected AppShellViewModel to be able to control menu items visibility via bound properties.

Trying to achieve:
![image](https://user-images.githubusercontent.com/64279914/230729767-07ab5889-8493-468e-bb8f-47c1d3ca26af.png)

With this code, resolving the resource failed with the error shown below:
```
public AppShell(AppShellViewModel viewModel)
    {
        InitializeComponent();
        this.BindingContext = viewModel;
    }
```
![image](https://user-images.githubusercontent.com/64279914/230729786-90385e66-0050-44a6-9b2d-3bcc0075508a.png)


With this code, resolving the resource works fine:
```
public App(IServiceProvider services)
    {
        InitializeComponent();
        var appShell = services.GetRequiredService<AppShell>();
        MainPage = appShell;
    }
```

**Comment by ToolmakerSteve** [comment]
[This StackOverflow question](https://stackoverflow.com/q/76369228/199364) is another example of the problem.

As I explain in first comment there, the problem is a direct consequence of how DI works.

The assumption in XamarinForms/Maui has always been that App gets initialized, BEFORE any page is created. Using DI to inject a page into App constructor breaks that design assumption.

**Comment by MSicc** [comment]
Having the same problem. 

- using DynamicResource removes the Exception, but does not resolve the Resource at runtime
- initializing the App.Current.MainPage with the injected View (not AppShell in my case, but a TabbedPage) does not work as well

**Comment by Magic73** [comment]
I hit this issue also. I fixed in this way:

`    public partial class App : Application
    {
        public App(IServiceProvider serviceProvider)
        {
            InitializeComponent();

            MainPage = serviceProvider.GetRequiredService<MainPage>();
        }
    }`

**Comment by dennis-garavsky** [comment]
I experienced the same behavior ("StaticResource not found for key ShellMenuItemTemplate") when injecting a custom service into my MainPage (derived from Shell) as part of my pet project powered by DevExpress .NET MAUI project templates and controls. Your solution certainly worked (`MainPage = serviceProvider.GetRequiredService<MainPage>();`) and it is fine with me - thank you for documenting it here for others.

Please just consider the following documentation enhancements for the .NET MAUI developer community (not urgent):
1. This DI specificity with StaticResource is not mentioned anywhere in the .NET MAUI docs (
https://learn.microsoft.com/en-us/dotnet/architecture/maui/dependency-injection), where the problematic approach in the App.xaml.cs is shown everywhere (`MainPage = mainPage;`) - probably your tech writers can add a note there. CC @Eilon 
2. I also feel that just like myself, many people searched the internet for prebuilt code samples using  [.NET MAUI + DI](https://www.google.com/search?q=maui+di), "maui dependency injection", and other relevant keywords). As a result, the following highly trustworthy resources from the recognized influencers demonstrated the same approach without highlighting this potential specificity - I am mentioning you directly just in case you may want to consider updates in the future.
CC @jamesmontemagno (https://montemagno.com/tag/dependency-injection/ and https://youtu.be/ugem4UbAtC0)
@jfversluis (https://youtu.be/wkgbvMlrMhU)

Thank you, Everyone.

**Comment by Zhanglirong-Winnie** [comment]
Verified this issue with Visual Studio Enterprise 17.8.0 Preview 3.0. Can repro on android platform with sample project.
[SampleProject-master.zip](https://github.com/dotnet/maui/files/13067967/SampleProject-master.zip)
![image](https://github.com/dotnet/maui/assets/77434865/49a3155a-d2b2-44ff-87bb-c6ba625ec843)

**Comment by thomasgalliker** [comment]
We cannot inject the "MainPage" into App ctor if MainPage uses static resources from App.xaml (or further references xamls, e.g. Styles.xaml or Colors.xaml). The instance of MainPage is created by DI just before it is injected into App.xaml.cs' constructor. We need to run App.xaml.cs' InitializeComponent() method before we can run MainPage's InitializeComponent() method.

I would expect .NET MAUI to provide better support for Page/ViewModel handling with DI.

**Comment by awaescher** [comment]
This issue was reported a whole year ago and noone from the team reacted so far.
StaticResources are a fundamental thing for ANY real world app that is not just another Maui Sample. Seriously, you can't be ignoring this issue every single enterprise customer is dependent on.

**Comment by Eilon** [comment]
Hi, I'm from the team 😁 

Here's what I've understood from this thread:

Using .NET MAUI + XAML + StaticResources + DI for pages causes an exception. To avoid the exception, the code can be rewritten to take in the `IServiceProvider` in the constructor (instead of taking in the page), and then the service provider can be asked to retrieve the page _after_ `InitializeComponent()` is called. This workaround has been posted earlier in this thread, and can be summarized like this:

**Causes exception:** (`Microsoft.Maui.Controls.Xaml.XamlParseException: 'Position row:column. StaticResource not found for key MyStaticResource'`)

```c#
	public App(LoginPage page)
	{
		InitializeComponent();
		MainPage = page;
	}
```

**Works:**

```c#
	public App(IServiceProvider sp)
	{
		InitializeComponent();
		MainPage = sp.GetService<LoginPage>();
	}
```

The cause of this is that XAML creates object trees *outside-in*, meaning the outer-most container is created first (say, the app container), then the next deeper one (say, a page or layout), etc., until finally the inner-most content is created (say, a button). However, Microsoft.Extensions.DependencyInjection creates object trees *inside-out*, meaning the inner-most item is created first, then whatever contains it is created next, until eventually the thing you 'asked' for is finally created.

This conflict is problematic because of the two different patterns: outside-in vs. inside-out.

The reason the workaround works is that it waits to do the DI part until after the XAML part is sufficiently done.

Given this fairly fundamental conflict, I'd like to hear on suggestions for how we can make this better.

1. Actually fix the problem. I'm not sure if that's even possible without some huge changes. If someone here thinks it _is_ possible, please share your thoughts on how we could do that. Cost: I'm not sure how expensive this is because I'm not sure if it's even possible.
1. Add docs. This was suggested earlier and I logged https://github.com/dotnet/docs-maui/issues/1919 with some ideas. Please add any additional thoughts on docs to that. Cost: This is relatively inexpensive.
1. Add a Code Analyzer to detect the pattern and at least warn about the problem, and possibly suggest a code fix for the user. Cost: This could be expensive, but I'm not sure, because I've never written a proper analyzer.

If anyone has some additional thoughts on how we can improve this, please let us know.

**Comment by symbiogenesis** [comment]
I think using a fully MVVM and DI-centric approach in the MAUI project templates would ensure people don't run into this.

IServiceProvider would be used by default to grab the initial MainPage and the ViewModel for the AppShell.

**Comment by awaescher** [comment]
I could not get the workaround to work for me. Therefore, I have no choice but to remove all static resources in order to continue my migration efforts. Might be another issue in my migrated codebase that is causing this in combination but as for almost every error I encounter, the error messages and callstacks are completely useless.

This migration from a medium sized enterprise Xamarin.Forms app was the worst experience I've ever had as a developer and that truly means a lot. I used to be a Xamarin and especially Xamarin.Forms advocate. I have been in from the very beginning and gladly paying $999 yearly for Xamarin.iOS.

These days I'm downright frustrated with how tedious the porting and how immature the runtime is. It's a disastrous state and the way I see it, nothing has really changed with .NET 8.

**Comment by janseris** [comment]
> > I could not get the workaround to work for me. Therefore, I have no choice but to remove all static resources in order to continue my migration efforts. Might be another issue in my migrated codebase that is causing this in combination but as for almost every error I encounter, the error messages and callstacks are completely useless.
> > This migration from a medium sized enterprise Xamarin.Forms app was the worst experience I've ever had as a developer and that truly means a lot. I used to be a Xamarin and especially Xamarin.Forms advocate. I have been in from the very beginning and gladly paying $999 yearly for Xamarin.iOS.
> > These days I'm downright frustrated with how tedious the porting and how immature the runtime is. It's a disastrous state and the way I see it, nothing has really changed with .NET 8.
> 
> Hi @awaescher if you can provide a simplified repro that shows the issue you're having I can try to investigate. My familiarity is mostly with the DI side of things and much less so on the XAML/resources side. I wonder in your case are there multiple levels of XAML pages using static resources? If so, the workaround I mentioned in my comment would need to be applied at each level. Or, it could be something else entirely.
> 
> There could also be other completely different approaches, such as querying each service "on demand" once the page loads, instead of using constructor injection at all. While this would most likely work, because it allows all the XAML stuff to work "normally," in some apps it could require significant changes to allow some functionality to happen after the page loads.

So we should use service locator antipattern because MAUI team cannot implement DI properly? No thanks

**Comment by Rabosa616** [comment]
I have a complex app with more than 30 pages and a bunch of services and the only think I´ve did was on the App.xaml.cs
at the constructor apply @Eilon workarround, then I´ve added all my static resources at App.xaml at:
```
 <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
```

Then on each page I use propertly the DI.
Here by an example:

MauiProgram.cs:

```csharp
public static class MauiProgram
{
	public static MauiApp CreateMauiApp()
	{
		var builder = MauiApp.CreateBuilder();
		builder
			.UseMauiApp<App>()
			.ConfigureFonts(fonts =>
			{
				fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
				fonts.AddFont("OpenSans-Semibold.ttf", "OpenSansSemibold");
			});
                        # register services
                        builder.Services.AddSingleton<IConnectivity>((e) => Connectivity.Current);

                        # register pages
                        builder.Services.AddSingleton<MainPage>();

                        # register view models
                        builder.Services.AddSingleton<MainPageViewModel>();

		return builder.Build();
	}
}
```

App.xaml
```xml
﻿<?xml version = "1.0" encoding = "UTF-8" ?>
<Application xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:local="clr-namespace:MauiApp1"
             x:Class="MauiApp1.App">
    <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <ResourceDictionary Source="Resources/Styles/Colors.xaml" />
                <ResourceDictionary Source="Resources/Styles/Styles.xaml" />
                 <!-- Add here your Resource Dictionaries -->
            </ResourceDictionary.MergedDictionaries>
        </ResourceDictionary>
    </Application.Resources>
</Application>
```

App.xaml.cs

```csharp
public partial class App : Application
    {
        public App(IServiceProvider sp)
        {
            InitializeComponent();

            MainPage = MainPage = sp.GetService<MainPage>();
        }
    }
```

MainPage.xaml.cs
```csharp
public partial class MainPage : ContentPage
{
    int count = 0;

    public MainPage(MainPageViewModel viewModel)
    {
        InitializeComponent();
        BindingContext = viewModel;
    }
   ...
```

MainPageViewModel.cs
```csharp
 public class MainPageViewModel
 {
     private readonly IConnectivity connectivityService
     public MainPageViewModel(IConnectivity connectivity)
     {
         connectivityService = connectivity ?? throw new ArgumentNullException(nameof(connectivity));
     }
 }
```

**Comment by awaescher** [comment]
>  can provide a simplified repro that shows the issue you're having I can try to investigate

Hard to provide as this is a larger app I am currently "upgrading" (not sure about that) from Xamarin.Forms to Maui. Because all the efforts with the upgrade-assistant went directly to hell, I started a new Maui app and manually moved class for class from my older project to the newer one. So it's not that I threw in loads of old tech, manifests and plists. Still, I just cannot get this thing to work, but without knowing why, it's hard to build a repro.

My code however looks very much like the one @Rabosa616 just posted. However, the troublemakers in my case are static resources in the XAML itself:

``` xaml
<?xml version="1.0" encoding="UTF-8"?>
<ContentPage
	x:Name="anyPage"
	x:Class="MyApp.Pages.AnyPage"
	Title="{StaticResource AppResources.MyAppName}">
	<ContentPage.Content>
		...
```

I just found this closed issue #3878 which is about the same thing, there's a repro repository.

**Comment by awaescher** [comment]
Okay wait, I am very confused to report that this works:

```csharp
Title="{Static AppResources.MyAppName}">
```

While this does not:

```csharp
Title="{StaticResource AppResources.MyAppName}">
```

```
Exception has occurred: Microsoft.Maui.Controls.Xaml.XamlParseException
Position 15:2. StaticResource not found for key AppResources.MyAppName
    at at Microsoft.Maui.Controls.Xaml.ApplyPropertiesVisitor.ProvideValue(Object& value, ElementNode node, Object source, XmlName propertyName)
    at at Microsoft.Maui.Controls.Xaml.ApplyPropertiesVisitor.Visit(ElementNode node, INode parentNode)
    at at Microsoft.Maui.Controls.Xaml.ElementNode.Accept(IXamlNodeVisitor visitor, INode parentNode)
    at at Microsoft.Maui.Controls.Xaml.RootNode.Accept(IXamlNodeVisitor visitor, INode parentNode)
    at at Microsoft.Maui.Controls.Xaml.XamlLoader.Visit(RootNode rootnode, HydrationContext visitorContext, Boolean useDesignProperties)
    at at Microsoft.Maui.Controls.Xaml.XamlLoader.Load(Object view, String xaml, Assembly rootAssembly, Boolean useDesignProperties)
    at at Microsoft.Maui.Controls.Xaml.XamlLoader.Load(Object view, String xaml, Boolean useDesignProperties)
    at at Microsoft.Maui.Controls.Xaml.XamlLoader.Load(Object view, Type callingType)
    at at Microsoft.Maui.Controls.Xaml.Extensions.LoadFromXaml[LandingPage](LandingPage view, Type callingType)
    at MyApp.Pages.AnyPage.InitializeComponent() in 
```

I am not sure if this is the way it should be when working when using resx files for localization, I used `StaticResource` with Xamarin.Forms over the last years.

**Comment by AndreasReitberger** [comment]
Just for reference, for styles and colors, which are defined in a `ResourceDictionary`, I use `{StaticResource ...}`, for classes and static resources from code behind, I'm using `{x:Static ...}`.
Haven't had any problems yet, even with `DI`.

```xaml
 <Label 
        Margin="2,16"
        Text="This is a headline"
        VerticalTextAlignment="Center" 
        HorizontalTextAlignment="Center"
        Style="{StaticResource PrimaryHeadlineLabelStyle}"
        />

    <Label 
        Margin="2,16"
        Text="{x:Static icons:MaterialIcons.HeartOutline}"
        VerticalTextAlignment="Center" 
        HorizontalTextAlignment="Center"
        Style="{StaticResource PrimaryMaterialFontFamilyIconLabelStyle}"
        />

```

The `MaterialFontIcons` class looks like that.

https://github.com/AndreasReitberger/SharedMauiXamlStyles/blob/main/src/SharedMauiXamlStylesLibrary/Models/FontIcons/MaterialIcons.cs

Maybe this helps!

**Comment by ThysBrits** [comment]
I was faced with this issue the last week and struggled to find any solution that worked. None of the above possible solutions worked for me. Then eventually I found that I had added a Style property on some of my controls that was invalid. The style was in the styles.xaml file, but I was using in incorrectly. And the error was clear enough that the style was not found. Once I removed those Style properties, the app worked using the DI, etc. The only bug though is the fact that the app would run when attached to the debugger, but it would crash when not. The way I managed to solve it better was to run the Windows app version, which does in fact give the error when attached to the debugger. Hope this helps someone, and also helps in resolving the actual problem.

**Comment by janseris** [comment]
> I was faced with this issue the last week and struggled to find any solution that worked. None of the above possible solutions worked for me. Then eventually I found that I had added a Style property on some of my controls that was invalid. The style was in the styles.xaml file, but I was using it incorrectly. And the error was clear enough that the style was not found. Once I removed those Style properties, the app worked using the DI, etc. The only bug though is the fact that the app would run when attached to the debugger, but it would crash when not. The way I managed to solve it better was to run the Windows app version, which does in fact give the error when attached to the debugger. Hope this helps someone, and also helps in resolving the actual problem.

the cause debug × release might be interpreted vs compiled XAML what I read in this repository sometimes.

**Comment by tele-bird** [comment]
I note that @jfversluis 's [DI injection sample project](https://github.com/jfversluis/MauiDependencyInjectionSample) injects a MainPage into the App constructor, and it runs without this issue.... even in [my pull request that upgrades it to .NET8](https://github.com/jfversluis/MauiDependencyInjectionSample/pull/2).

**Comment by janseris** [comment]
btw proper DI (everything in the app resolved from DI) allows developers to validate DI setup on application start.
If anything is resolved later during runtime, that cannot be done.

The code for doing that is:

```
private static void ValidateServiceDependenciesOnStartup(IHostBuilder builder)
        {
            builder.UseDefaultServiceProvider(options =>
            {
                options.ValidateOnBuild = true; //validate on builder.Build()
            });
        }
```

**Comment by cyraid** [comment]
I was having the same issues, but I was putting my `<Style>` and `<x:Double>` etc in `<Application.Resources>` directly, but I moved them into `<ResourceDictionary>` instead (inside Application.Resources) and it worked after.
